docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+  * Register the pod with other components of the system
+* Reduce coupling between application images and reduce the need to customize
+  those images for Kubernetes generally or specific roles
+* Reduce the footprint of individual images by specializing which containers
+  perform which tasks (install git into init container, use filesystem contents
+  in web container)
+* The order init containers start should be predictable and allow users to easily
+  reason about the startup of a container
+* Complex ordering and failure is out of scope - all complex workflows can if
+  necessary be implemented inside of a single init container, and this proposal
+  aims to enable that ordering without adding undue complexity to the system
+* Both run-once and run-forever pods should be able to use init containers
+* As much as possible, an init container should behave like an app container
+  to reduce complexity for end users, for clients, and for divergent use cases.
+  An init container is a container with the minimum alterations to accomplish
+  its goal.
+
+## Alternatives
+
+* Any mechanism that runs user code on a node before regular pod containers
+  should itself be a container and modeled as such - we explicitly reject
+  creating new mechanisms for running user processes.
+* The container pre-start hook (not yet implemented) requires execution within
+  the container's image and so cannot adapt existing images. It also cannot
+  block startup of containers
+* Running a "pre-pod" would defeat the purpose of the pod being an atomic
+  unit of scheduling.
+
+
+## Design
+
+Each pod may have 0..N init containers defined along with the existing
+1..N app containers.
+
+On startup of the pod, after the network and volumes are initialized, the
+init containers are started in order. Each container must exit successfully
+before the next is invoked. If a container fails to start, it is retried
+according to the pod RestartPolicy. RestartPolicyNever pods will immediately
+fail and exit. RestartPolicyAlways pods will retry the failing init container
+with increasing backoff until it succeeds. Future revisions of this spec may
+add the ability for an init container to define a limited container restart
+policy distinct from the pod RestartPolicy.
+
+A pod cannot be ready until all init containers have succeeded. The ports
+on an init container are not aggregated under a service. A pod that is
+being initialized is in the `Pending` phase but should have a distinct
+condition.
+
+If the pod is "restarted" all containers stopped and started due to
+a node restart, change to the pod definition, or admin interaction, all
+init containers must execute again.
+
+Each init container has all of the fields of an app container. The following
+fields are prohibited from being used on init containers by validation:
+
+* `readinessProbe` - init containers must exit for pod startup to continue,
```

Comment:
You could be starting a binary that runs an initialization process that has
a health check, or you could use an exec liveness probe to verify the last
touch date on the git repo is shorter than some interval, probably.

On Mon, Apr 18, 2016 at 5:38 PM, Phillip Wittrock <notifications@github.com>
wrote:

> In docs/proposals/container-init.md
> <https://github.com/kubernetes/kubernetes/pull/23666#discussion_r60138977>
> :
>
> > +add the ability for an init container to define a limited container restart
> > +policy distinct from the pod RestartPolicy.
> > +
> > +A pod cannot be ready until all init containers have succeeded. The ports
> > +on an init container are not aggregated under a service. A pod that is
> > +being initialized is in the `Pending` phase but should have a distinct
> > +condition.
> > +
> > +If the pod is "restarted" all containers stopped and started due to
> > +a node restart, change to the pod definition, or admin interaction, all
> > +init containers must execute again.
> > +
> > +Each init container has all of the fields of an app container. The following
> > +fields are prohibited from being used on init containers by validation:
> > +
> > +* `readinessProbe` - init containers must exit for pod startup to continue,
>
> It isn't immediately obvious to me how one would generally do liveness
> checking, since init containers would be run to completion commands. How do
> we check that git isn't hung? Do Http and Tcp liveness checks make sense in
> this context?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23666/files/7efdc3d36197f1ca452339d49abbc87163fb0e75#r60138977>
>